radix-engine-interface/src/blueprints/macros.rs

```diff
@@ -0,0 +1,145 @@
+macro_rules! define_invocation {
```

Comment:
fun! Can you add a comment that we should aim to converge this with the scrypto side over the long term?

Reply:
Sure thing, will do! I'm experimenting with a `native_blueprint_interface` macro that would do the following for us:

* Define the blueprint's functions ident constants
* Define the blueprint's functions export name constants
* Define the blueprint's functions input structs
* Define the blueprint's functions manifest input structs
* Define the blueprint's functions output structs
* Define the schema of the blueprint
* Define the stubs for the blueprint

So far the syntax for that looks like the following.

```rust
native_blueprint_interface! {
    struct Account {
        default_deposit_rule: AccountDefaultDepositRule,
    }

    // Functions & Methods

    fn create_local() -> Own;

    fn create() -> (ComponentAddress, Bucket);

    virtual fn create_virtual_ecdsa_secp256k1();

    virtual fn create_virtual_ecdsa_ed25519();

    fn create_advanced(authority_rules: AuthorityRules) -> ComponentAddress;

    fn securify(&mut self) -> Bucket;

    fn lock_fee(&mut self, amount: Decimal);

    fn lock_contingent_fee(&mut self, amount: Decimal);

    fn deposit(
        &mut self,
        #[interface(manifest_type = "ManifestBucket")]
        bucket: Bucket,
    );

    fn try_deposit_or_abort(
        &mut self,
        #[interface(manifest_type = "ManifestBucket")]
        bucket: Bucket,
    );

    fn try_deposit_or_refund(
        &mut self,
        #[interface(manifest_type = "ManifestBucket")]
        bucket: Bucket,
    ) -> Bucket;

    fn deposit_batch(
        &mut self,
        #[interface(manifest_type = "Vec<ManifestBucket>")]
        buckets: Vec<Bucket>,
    );

    fn try_deposit_batch_or_abort(
        &mut self,
        #[interface(manifest_type = "Vec<ManifestBucket>")]
        buckets: Vec<Bucket>,
    );

    fn try_deposit_batch_or_refund(
        &mut self,
        #[interface(manifest_type = "Vec<ManifestBucket>")]
        buckets: Vec<Bucket>,
    ) -> Vec<Bucket>;

    fn withdraw(
        &mut self,
        resource_address: ResourceAddress,
        amount: Decimal
    ) -> Bucket;

    fn withdraw_non_fungibles(
        &mut self,
        resource_address: ResourceAddress,
        ids: BTreeSet<NonFungibleLocalId>
    ) -> Bucket;

    fn lock_fee_and_withdraw(
        &mut self,
        amount_to_lock: Decimal,
        resource_address: ResourceAddress,
        amount: Decimal
    ) -> Bucket;

    fn lock_fee_and_withdraw_non_fungibles(
        &mut self,
        amount_to_lock: Decimal,
        resource_address: ResourceAddress,
        ids: BTreeSet<NonFungibleLocalId>
    ) -> Bucket;

    fn create_proof(
        &mut self,
        resource_address: ResourceAddress,
    ) -> Proof;

    fn create_proof_of_amount(
        &mut self,
        resource_address: ResourceAddress,
        amount: Decimal
    ) -> Proof;

    fn create_proof_of_non_fungibles(
        &mut self,
        resource_address: ResourceAddress,
        ids: BTreeSet<NonFungibleLocalId>
    ) -> Bucket;

    fn change_account_default_deposit_rule(
        &mut self,
        default_deposit_rule: AccountDefaultDepositRule
    ) -> Bucket;

    fn configure_resource_deposit_rule(
        &mut self,
        resource_address: ResourceAddress,
        resource_deposit_configuration: ResourceDepositRule
    );
}
```

So far the macro is able to the first five items, when I get some more time (if we like this approach) I can continue this and have it do this stuff for us.